File: scripts/lib/deduplicator.ts
import type { PartialDevelopment, ExtractedDevelopment, DevelopmentType, DevelopmentStatus } from './types.js';
import { generateSlug } from './utils.js';
import { scoreConfidence } from './confidence.js';

const SOURCE_PRIORITY: Record<string, number> = {
  'operator_website': 0,
  'property_portal': 1,
  'news': 2,
  'planning': 3,
  'other': 4,
};

export function deduplicateAndMerge(
  partials: PartialDevelopment[],
  operatorName: string,
  defaultType: DevelopmentType,
  operatorDomain?: string,
): ExtractedDevelopment[] {
  const groups = new Map<string, PartialDevelopment[]>();

  for (const dev of partials) {
    if (!dev.name || dev.name.length < 2) continue;

    const slug = generateSlug(dev.name);
    const matchKey = findMatchKey(groups, slug, dev.name);

    if (matchKey) {
      groups.get(matchKey)!.push(dev);
    } else {
      groups.set(slug, [dev]);
    }
  }

  // Resolve slug collisions (same slug, different areas)
  const slugCounts = new Map<string, number>();
  const results: ExtractedDevelopment[] = [];

  for (const [baseSlug, group] of groups) {
    const merged = mergeGroup(group, operatorName, defaultType);

    // Handle slug collision by appending area
    let finalSlug = baseSlug;
    const count = slugCounts.get(baseSlug) || 0;
    if (count > 0 && merged.area) {
      finalSlug = `${baseSlug}-${generateSlug(merged.area)}`;
    }
    slugCounts.set(baseSlug, count + 1);

    merged.slug = finalSlug;
    merged.confidence = scoreConfidence(merged, operatorDomain);
    results.push(merged);
  }

  // Sort by confidence (highest first)
  results.sort((a, b) => b.confidence.overall - a.confidence.overall);

  return results;
}

function findMatchKey(
  groups: Map<string, PartialDevelopment[]>,
  candidateSlug: string,
  candidateName: string
): string | null {
  // Exact slug match
  if (groups.has(candidateSlug)) return candidateSlug;

  // Fuzzy name match
  const normalizedCandidate = candidateName.toLowerCase().replace(/[^a-z0-9]/g, '');

  for (const [key, group] of groups) {
    const existingName = group[0].name.toLowerCase().replace(/[^a-z0-9]/g, '');

    // One contains the other
    if (
      normalizedCandidate.includes(existingName) ||
      existingName.includes(normalizedCandidate)
    ) {
      return key;
    }

    // Very similar slugs (differ by <=2 chars at the end, e.g. "the-forge" vs "the-forges")
    if (
      candidateSlug.length > 5 &&
      key.length > 5 &&
      (candidateSlug.startsWith(key) || key.startsWith(candidateSlug))
    ) {
      const diff = Math.abs(candidateSlug.length - key.length);
      if (diff <= 2) return key;
    }
  }

  return null;
}

function mergeGroup(
  group: PartialDevelopment[],
  operatorName: string,
  defaultType: DevelopmentType
): ExtractedDevelopment {
  // Sort by source priority (operator_website first)
  const sorted = [...group].sort(
    (a, b) => (SOURCE_PRIORITY[a.sourceType] ?? 4) - (SOURCE_PRIORITY[b.sourceType] ?? 4)
  );

  const sourceUrls = [...new Set(group.map(d => d.sourceUrl))];
  const notes: string[] = [];

  // Pick best value for each field
  const name = pickBestString(sorted.map(d => d.name))!;
  const area = pickBestString(sorted.map(d => d.area));
  const region = pickBestString(sorted.map(d => d.region));
  const postcode = pickBestString(sorted.map(d => d.postcode));
  const websiteUrl = pickBestString(sorted.map(d => d.websiteUrl));
  const description = pickBestString(sorted.map(d => d.description));
  const assetOwner = pickBestString(sorted.map(d => d.assetOwner));
  const completionDate = pickBestString(sorted.map(d => d.completionDate));

  // Pick best numeric value (prefer non-null from higher-priority source)
  const numberOfUnits = pickBestNumber(sorted.map(d => d.numberOfUnits));

  // Pick status (prefer explicit from higher-priority source)
  const status = pickBestStatus(sorted.map(d => d.status));

  // Pick development type
  const devType = pickBestDevType(sorted.map(d => d.developmentType), defaultType);

  // Merge amenities (OR across all sources)
  const amenities: Record<string, boolean> = {};
  for (const dev of sorted) {
    if (dev.amenities) {
      for (const [key, val] of Object.entries(dev.amenities)) {
        if (val) amenities[key] = true;
      }
    }
  }

  // Pets allowed (OR)
  const petsAllowed = sorted.some(d => d.petsAllowed);

  // Year completed
  let yearCompleted: number | null = null;
  if (completionDate) {
    const year = parseInt(completionDate.substring(0, 4));
    if (year >= 2015 && year <= 2035) yearCompleted = year;
  }

  // Add notes for data quality
  if (!numberOfUnits) notes.push('Unit count not confirmed');
  if (!status) notes.push('Status not confirmed');
  if (!postcode) notes.push('Postcode not found');
  if (!region) notes.push('Region could not be determined');
  if (group.length > 1) {
    const uniqueUnitCounts = [...new Set(group.map(d => d.numberOfUnits).filter(Boolean))];
    if (uniqueUnitCounts.length > 1) {
      notes.push(`Conflicting unit counts: ${uniqueUnitCounts.join(', ')}`);
    }
  }

  return {
    name,
    slug: '', // will be set by caller
    developmentType: devType,
    area: area || '',
    region: region || '',
    postcode: postcode || '',
    numberOfUnits: numberOfUnits ?? null,
    status: status ?? null,
    completionDate: completionDate || '',
    yearCompleted,
    description: description || '',
    websiteUrl: websiteUrl || '',
    assetOwner: assetOwner || operatorName,
    operator: operatorName,
    amenities,
    petsAllowed,
    confidence: {
      overall: 0,
      level: 'LOW',
      sourceCount: sourceUrls.length,
      sourceTypes: [],
      notes: [],
    },
    sourceUrls,
    extractionNotes: notes,
  };
}

function pickBestString(values: (string | undefined)[]): string | undefined {
  for (const v of values) {
    if (v && v.trim().length > 0) return v.trim();
  }
  return undefined;
}

function pickBestNumber(values: (number | undefined | null)[]): number | null {
  for (const v of values) {
    if (v !== undefined && v !== null && v > 0) return v;
  }
  return null;
}

function pickBestStatus(values: (DevelopmentStatus | undefined | null)[]): DevelopmentStatus | null {
  for (const v of values) {
    if (v) return v;
  }
  return null;
}

function pickBestDevType(values: (DevelopmentType | undefined)[], defaultType: DevelopmentType): DevelopmentType {
  for (const v of values) {
    if (v) return v;
  }
  return defaultType;
}

File: scripts/discover/deduplicator.py
import re
from urllib.parse import urlparse

from models import (
    Confidence,
    DiscoveredDevelopment,
    VALID_REGIONS,
    VALID_STATUSES,
)


def generate_slug(text: str) -> str:
    """Generate a URL-safe slug. Matches lib/utils.ts:generateSlug()."""
    slug = text.lower().strip()
    slug = re.sub(r"[^\w\s-]", "", slug)
    slug = re.sub(r"\s+", "-", slug)
    slug = re.sub(r"-+", "-", slug)
    return slug.strip("-")


def deduplicate_developments(raw_developments: list[dict]) -> list[DiscoveredDevelopment]:
    """
    Group raw LLM extractions by slug/fuzzy name, merge duplicates,
    score confidence, and return a sorted list of DiscoveredDevelopment.
    """
    groups: dict[str, list[dict]] = {}

    for dev in raw_developments:
        name = dev.get("name", "").strip()
        if not name or len(name) < 3:
            continue

        slug = generate_slug(name)
        if not slug:
            continue

        match_key = _find_match_key(groups, slug, name)
        if match_key:
            groups[match_key].append(dev)
        else:
            groups[slug] = [dev]

    results = []
    for slug, group in groups.items():
        merged = _merge_group(slug, group)
        merged.confidence_score = _score_confidence(merged)
        if merged.confidence_score >= 0.7:
            merged.confidence = Confidence.HIGH
        elif merged.confidence_score >= 0.4:
            merged.confidence = Confidence.MEDIUM
        else:
            merged.confidence = Confidence.LOW
        results.append(merged)

    results.sort(key=lambda d: d.confidence_score, reverse=True)
    return results


def _find_match_key(groups: dict[str, list[dict]], candidate_slug: str, candidate_name: str) -> str | None:
    """Fuzzy slug/name matching to group duplicate extractions."""
    if candidate_slug in groups:
        return candidate_slug

    normalized = re.sub(r"[^a-z0-9]", "", candidate_name.lower())
    for key, group in groups.items():
        existing_name = re.sub(r"[^a-z0-9]", "", group[0].get("name", "").lower())

        # Substring containment
        if len(normalized) > 5 and len(existing_name) > 5:
            if normalized in existing_name or existing_name in normalized:
                return key

        # Slug prefix match (differ by <=2 chars)
        if len(candidate_slug) > 5 and len(key) > 5:
            if candidate_slug.startswith(key) or key.startswith(candidate_slug):
                if abs(len(candidate_slug) - len(key)) <= 2:
                    return key

    return None


def _merge_group(slug: str, group: list[dict]) -> DiscoveredDevelopment:
    """Merge multiple extractions of the same development into one record."""
    # Collect all source URLs
    source_urls = []
    seen_sources = set()
    for dev in group:
        src = dev.get("_source_url", "")
        if src and src not in seen_sources:
            source_urls.append(src)
            seen_sources.add(src)

    # Pick best value for each field (first non-empty wins)
    def pick(field: str) -> str | None:
        for dev in group:
            val = dev.get(field)
            if val is not None and str(val).strip():
                return str(val).strip()
        return None

    def pick_int(field: str) -> int | None:
        for dev in group:
            val = dev.get(field)
            if val is not None:
                try:
                    return int(val)
                except (ValueError, TypeError):
                    continue
        return None

    name = pick("name") or group[0].get("name", "Unknown")
    region = pick("region")
    status = pick("status")

    # Validate constrained fields
    if region and region not in VALID_REGIONS:
        region = None
    if status and status not in VALID_STATUSES:
        status = None

    dev_type = pick("development_type") or "Multifamily"
    if dev_type not in ("Multifamily", "Single Family"):
        dev_type = "Multifamily"

    return DiscoveredDevelopment(
        name=name,
        slug=slug,
        development_type=dev_type,
        operator_name=pick("operator_name"),
        asset_owner_name=pick("asset_owner_name"),
        area=pick("area"),
        region=region,
        postcode=pick("postcode"),
        number_of_units=pick_int("number_of_units"),
        status=status,
        completion_date=pick("completion_date"),
        description=pick("description"),
        website_url=pick("website_url"),
        source_urls=source_urls,
    )


def _score_confidence(dev: DiscoveredDevelopment) -> float:
    """Score confidence 0.0-1.0 based on available data and source count."""
    score = 0.0

    # Name (always present)
    if dev.name and len(dev.name) > 2:
        score += 0.15

    # Location fields
    if dev.postcode:
        score += 0.10
    if dev.area:
        score += 0.05
    if dev.region:
        score += 0.05

    # Core details
    if dev.number_of_units:
        score += 0.10
    if dev.status:
        score += 0.07
    if dev.operator_name:
        score += 0.08

    # Source count bonus
    src_count = len(dev.source_urls)
    if src_count >= 3:
        score += 0.20
    elif src_count >= 2:
        score += 0.15
    elif src_count >= 1:
        score += 0.05

    # Source type bonus
    for url in dev.source_urls:
        src_type = _classify_source(url)
        if src_type == "news":
            score += 0.02
        elif src_type == "property_portal":
            score += 0.03

    return min(score, 1.0)


NEWS_SITES = [
    "btrnews.co.uk", "reactnews.com", "propertyweek.com",
    "egi.co.uk", "estatesgazette.com", "insidehousing.co.uk",
    "placenorth.co.uk", "urbanliving.news",
]

PROPERTY_PORTALS = [
    "rightmove.co.uk", "zoopla.co.uk", "onthemarket.com",
    "openrent.com", "spareroom.co.uk",
]

PLANNING_SITES = [
    "planningportal.co.uk", "planningresource.co.uk",
]


def _classify_source(url: str) -> str:
    """Classify a URL by source type."""
    try:
        domain = urlparse(url).netloc.lower().replace("www.", "")
    except Exception:
        return "other"

    if any(site in domain for site in NEWS_SITES):
        return "news"
    if any(site in domain for site in PROPERTY_PORTALS):
        return "property_portal"
    if any(site in domain for site in PLANNING_SITES):
        return "planning"
    return "other"
